lib/internal/url.js

```diff
@@ -0,0 +1,328 @@
+    binding.parse(input.trim(), -1, base_context,
+      (flags, protocol, username, password,
+       host, port, path, query, fragment) => {
+        if (flags & binding.URL_FLAGS_FAILED)
+          throw new Error('Invalid URL');
+        this[context] = new StorageObject();
+        this[context].flags = flags;
+        this[context].scheme = protocol;
+        this[context].username = username;
+        this[context].password = password;
+        this[context].port = port;
+        this[context].path = path;
+        this[context].query = query;
+        this[context].fragment = fragment;
+        if (host !== undefined)
+          this[context].host = punycode.toASCII(host);
+      });
+  }
+
+  get [cannotBeBase]() {
+    return (this[context].flags & binding.URL_FLAGS_CANNOT_BE_BASE) != 0;
+  }
+
+  get protocol() {
+    return this[context].scheme;
+  }
+
+  set protocol(scheme) {
+    if (typeof scheme !== 'string')
+      throw new TypeError('protocol must be a string');
+    binding.parse(scheme, binding.kScheme, null, (...args) => {
+      // TODO: handle failure cases
+      this[context].scheme = args[binding.ARG_PROTOCOL];
+    });
+  }
+
+  get username() {
+    return this[context].username;
+  }
+
+  set username(username) {
+    if (typeof username !== 'string')
+      throw new TypeError('username must be a string');
+    this[context].username = encodeAuth(username);
+  }
+
+  get password() {
+    return this[context].password;
+  }
+
+  set password(password) {
+    if (typeof password !== 'string')
+      throw new TypeError('password must be a string');
+    this[context].password = encodeAuth(password);
+  }
+
+  get host() {
+    var ret = this[context].host;
+    if (this[context].port !== undefined)
+      ret += `:${this[context].port}`;
+    return ret;
+  }
+
+  set host(host) {
+    if (typeof host !== 'string')
+      throw new TypeError('host must be a string');
+    if (this[cannotBeBase])
+      return;
+    binding.parse(host, binding.kHost, null, (...args) => {
+      // TODO: handle failure cases
+      this[context].host = args[binding.ARG_HOST];
+      if (args[binding.ARG_PORT] !== undefined)
+        this[context].port = args[binding.ARG_PORT];
+    });
+  }
+
+  get hostname() {
+    return this[context].host;
+  }
+
+  set hostname(host) {
+    if (typeof host !== 'string')
+      throw new TypeError('host must be a string');
+    if (this[cannotBeBase])
+      return;
+    binding.parse(host, binding.kHostname, null, (...args) => {
+      // TODO: handle failure cases
+      this[context].host = args[binding.ARG_HOST];
+    });
+  }
+
+  get port() {
+    return this[context].port;
+  }
+
+  set port(port) {
+    if (!this[context].host || this[cannotBeBase] || this.protocol === 'file:')
+      return;
+    port |= 0;
+    if (port < 0 || port > 0xffff)
+      throw new RangeError('port must be a number between 0 and 65535');
+    this[context].port = port;
```

Comment:
This seems very not correct per spec: https://url.spec.whatwg.org/#port-state. Are you running the setter tests?